concordium-consensus/src/Concordium/GlobalState/Persistent/Account/CooldownQueue.hs

```diff
@@ -0,0 +1,215 @@
+                  prePreCooldown = Present target
+                }
+
+-- | Migrate a cooldown queue unchanged.
+migrateCooldownQueue :: forall m t av. (SupportMigration m t) => CooldownQueue av -> t m (CooldownQueue av)
+migrateCooldownQueue EmptyCooldownQueue = return EmptyCooldownQueue
+migrateCooldownQueue (CooldownQueue queueRef) =
+    CooldownQueue <$> migrateEagerBufferedRef return queueRef
+
+-- | Get the total stake in cooldown, pre-cooldown and pre-pre-cooldown.
+cooldownStake :: CooldownQueue av -> Amount
+cooldownStake EmptyCooldownQueue = 0
+cooldownStake (CooldownQueue queueRef) = cooldownTotal $ eagerBufferedDeref queueRef
+
+-- | Add the given amount to the pre-pre-cooldown.
+addPrePreCooldown ::
+    (MonadBlobStore m, AVSupportsFlexibleCooldown av) =>
+    -- | The amount to add to the pre-pre-cooldown.
+    Amount ->
+    CooldownQueue av ->
+    m (CooldownQueue av)
+addPrePreCooldown amount EmptyCooldownQueue = initialPrePreCooldownQueue amount
+addPrePreCooldown amount (CooldownQueue queueRef) = do
+    let oldCooldowns = eagerBufferedDeref queueRef
+    let !newCooldowns = Cooldowns.addPrePreCooldown amount oldCooldowns
+    makeCooldownQueue newCooldowns
+
+-- | Remove up to the given amount from the cooldowns, starting with pre-pre-cooldown, then
+--  pre-cooldown, and finally from the amounts in cooldown, in decreasing order of timestamp.
+reactivateCooldownAmount ::
+    (MonadBlobStore m, AVSupportsFlexibleCooldown av) =>
+    -- | The amount to reactivate.
+    Amount ->
+    CooldownQueue av ->
+    m (CooldownQueue av)
+reactivateCooldownAmount _ EmptyCooldownQueue = return EmptyCooldownQueue
+reactivateCooldownAmount amount (CooldownQueue queueRef) = do
+    let oldCooldowns = eagerBufferedDeref queueRef
+    let !newCooldowns = Cooldowns.reactivateCooldownAmount amount oldCooldowns
+    makeCooldownQueue newCooldowns
+
+-- | Process all cooldowns that expire at or before the given timestamp.
+--   This returns the next timestamp at which a cooldown expires, if any.
+processCooldownsUntil ::
+    (MonadBlobStore m) =>
+    -- | Release all cooldowns up to and including this timestamp.
+    Timestamp ->
+    CooldownQueue av ->
+    m (Maybe Timestamp, CooldownQueue av)
+processCooldownsUntil _ EmptyCooldownQueue = return (Nothing, EmptyCooldownQueue)
+processCooldownsUntil ts (CooldownQueue queueRef) = do
+    let !newCooldowns = processCooldowns ts $ eagerBufferedDeref queueRef
+    let !nextTimestamp = firstCooldownTimestamp newCooldowns
+    newQueue <- makeCooldownQueue newCooldowns
+    return (nextTimestamp, newQueue)
+
+-- | Move the pre-cooldown amount on into cooldown with the specified release time.
+--  This returns @Just (Just ts)@ if the previous next cooldown time was @ts@, but the new next
+--  cooldown (i.e. the supplied timestamp) time is earlier. It returns @Just Nothing@ if there was
+--  no cooldown but now there is. Otherwise, it returns @Nothing@.
+processPreCooldown ::
```

Comment:
This would benefit from return type with constructors for clear semantics, i.e. something like `NewCooldownInserted` instead of `Just Nothing`.